### ISSUE TITLE
Adapt WaveDiff training to account for masks

### ISSUE THREAD

**Issue by CentofantiEze** [CosmoStat/wf-psf#143]
Real data will contain masks for cosmic rays so we want to avoid using this pixels for computing the loss and back-propagating the gradients.

I suggest starting with a (wavediff) simulated dataset and incorporate either simulated or real masks.

Validation TBD. One idea is verifying that training without masks or training with masks full of ones give the same results. 

Team: @tobias-liaudat @jeipollack @CentofantiEze

**Comment by jeipollack** [comment]
Thanks @CentofantiEze for helping out with this task.

@tobias-liaudat  I sent Ezequiel the same dataset sample I sent to you.  Did you start a branch already that maybe he could continue?

**Comment by tobias-liaudat** [comment]
Thanks both! I haven't opened a branch yet, we should coordinate

**Comment by jeipollack** [comment]
I am available tomorrow or M-W next week.

**Comment by tobias-liaudat** [comment]
Could we do Monday next week? Morning or early afternoon?

**Comment by CentofantiEze** [comment]
I prefer Monday morning since I have a meeting at 13:30.

**Comment by jeipollack** [comment]
Monday morning works for me.

**Comment by tobias-liaudat** [comment]
Great, is 10am good?

**Comment by jeipollack** [comment]
Just add it would be worth looking at the effect the masks have on the centroid correction with simulated stars.

I have to check if there are unit tests for Tobias' centroid module, but I already started to write some tests in a notebook that could be used as unit and/or validation tests.

**Comment by tobias-liaudat** [comment]
1. Try to implement the masks and the weights in the likelihood.

2. Take into account the mask when:
    - Computing the noise std dev
    - Compute the centroid estimation

3. Consider if we drop stars with masking in the core or try to take them into account by adapting the centroid estimation:
    - It depends on how conservative we want to be.

**Comment by CentofantiEze** [comment]
I created a custom loss for handling the masks and validated the approach by training WaveDiff on masked data.

Results:

<img width="1122" alt="Image" src="https://github.com/user-attachments/assets/cdec1f69-1205-482c-92a4-d2c1f5c24359" />

In the attached figure you can see the test metrics (which doesn't include the mask yet) for different models.
- Control Train: baseline trained and tested on the gh repo dataset.
- Control Mask Test: trained on the gh dataset, tested on masked stars without using masks (to see the amount of error we expect)
- Mask train mask test: we train and test on masked data (error matches previous case since masks are not considered for the test metrics)
- Mask train unit mask test: train on masked data using masks, test on unmasked data or unit masked data (shows the model has trained properly on masked data)
- Unitary masks: train and test on data masked with unitary masks (serve as a control case for the mask handling, same results as control)

**Comment by CentofantiEze** [comment]
The new masked-mse loss overrides the predefined `__call__` method which by default forces `y_true` and `y_pred` to have matching shapes.

In the new masked loss `y_pred` has shape `(batch, output_dim, output_dim)` and `y_true` contains both the target stars (which are already masked) and the masks stacked over the last dimension, having shape `(batch, output_dim, output_dim, 2)`.

The target and masks are stacked in the train script and passed as the `outputs` variable to the `general_train_cycle()` function.

**Comment by CentofantiEze** [comment]
Account for sample weights in the new masked loss and metric.

The masked loss code is the following:

```python
def masked_mse(y_true, y_pred, mask, sample_weight=None):
    """Masked Mean Squared Error.

    Parameters
    ----------
    y_true: Tensor
        True values
    y_pred: Tensor
        Predicted values
    mask: Tensor
        Mask to be applied
    sample_weight: Tensor
        Sample weights

    Returns
    -------
    Tensor
        Masked Mean Squared Error
    """
    # Calculate the masked squared error
    error = mask * tf.square(y_true - y_pred) # (batch, height, width)
    # Apply sample weights if provided
    if sample_weight is not None:
        error *= tf.reshape(sample_weight, (-1, 1, 1))
    unmasked_pixels = tf.reduce_sum(mask, axis=[1, 2]) # (batch,)
    # Weight by number of non masked pixels
    return tf.reduce_sum(error / tf.reshape(unmasked_pixels, (-1, 1, 1)))

class MaskedMeanSquaredError(tf.keras.losses.Loss):
    """Masked Mean Squared Error.""" 
    def __init__(self, name="masked_mean_squared_error", **kwargs):
        super().__init__(name=name, **kwargs)

    def __call__(self, y_true, y_pred, sample_weight=None):
        """Overrides __call__() to allow y_true and y_pred to have different shapes."""
        return self.call(y_true, y_pred, sample_weight)

    def call(self, y_true, y_pred, sample_weight=None):
        """Calculates the loss."""
        # Extract the target and the masks from y_true
        y_target = y_true[..., 0]
        mask = y_true[..., 1]
        return masked_mse(y_target, y_pred, mask, sample_weight)
```

**Comment by CentofantiEze** [comment]
Modified `metrics.py` such that the masks are being considered for computing the pixel polychromatic low-res metrics.

**Comment by CentofantiEze** [comment]
Open Draft Pull Request #142.

**Comment by CentofantiEze** [comment]
Final results for masked loss training and testing validation.

<img width="900" alt="Image" src="https://github.com/user-attachments/assets/ce36bdf4-09f8-4379-a61f-0f665c568131" />
